docs/fundamentals/code-analysis/quality-rules/ca1846.md

```diff
@@ -0,0 +1,60 @@
+---
+title: "CA1846: Prefer `AsSpan` over `Substring`"
+description: "Learn about code analysis rule CA1846: Prefer `AsSpan` over `Substring`"
+ms.date: 05/21/2021
+ms.topic: reference
+f1_keywords:
+- CA1846
+- PreferAsSpanOverSubstring
+helpviewer_keywords:
+- PreferAsSpanOverSubstring
+- CA1846
+author: NewellClark
+dev_langs:
+- CSharp
+- VB
+---
+# CA1846: Prefer `AsSpan` over `Substring`
+
+| | Value |
+|-|-|
+| **Rule ID** |CA1846|
+| **Category** |[Performance](performance-warnings.md)|
+| **Fix is breaking or non-breaking** |Non-breaking|
+
+## Cause
+
+The result of a call to <xref:System.String.Substring(System.Int32)> or <xref:System.String.Substring(System.Int32, System.Int32)> is passed to a method with an available overload that accepts <xref:System.ReadOnlySpan{System.Char}>.
+
+## Rule description
+
+`Substring` allocates a new `string` object on the heap and performs a full copy of the extracted text. String manipulation is a performance bottleneck for many programs. Allocating many small short-lived strings on a hot path can create enough collection pressure to impact performance. The O(n) copies created by `Substring` become relevant when the substrings get large. <xref:System.Span%601> and <xref:System.ReadOnlySpan%601> were created to solve these performance problems.
+
+Many APIs that accept strings provide overloads that accept <xref:System.ReadOnlySpan{System.Char}>. When such overloads are available, you can improve performance by calling `AsSpan` instead of `Substring`.
```

Comment:
```suggestion
Many APIs that accept strings also have overloads that accept a `ReadOnlySpan<System.Char>` argument. When such overloads are available, you can improve performance by calling `AsSpan` instead of `Substring`.
```